tests/chainer_tests/optimizer_hooks_tests/test_gradient_clipping.py

```diff
@@ -1,110 +1,92 @@
+]
 
 
 class SimpleLink(chainer.Link):
 
-    def __init__(self, w, g):
+    def __init__(self, params):
         super(SimpleLink, self).__init__()
         with self.init_scope():
-            self.param = chainer.Parameter(I.Zero(), w.shape)
-        self.param.data = w
-        self.param.grad = g
-
-
-class TestOptimizerUtility(unittest.TestCase):
-
-    def setUp(self):
-        self.x = np.linspace(-1.0, 1.5, num=6).astype(np.float32).reshape(2, 3)
-        self.a = np.array(2.0)
-
-    def test_sqnorm_cpu(self):
-        # \Sum_{n=0}^{5} (-1.0+0.5n)**2 = 4.75
-        self.assertAlmostEqual(gradient_clipping._sum_sqnorm([self.x]), 4.75)
-
-    def test_sqnorm_scalar_cpu(self):
-        self.assertAlmostEqual(gradient_clipping._sum_sqnorm([self.a]), 4)
-
-    @attr.gpu
-    def test_sqnorm_gpu(self):
-        x = cuda.to_gpu(self.x)
-        self.assertAlmostEqual(gradient_clipping._sum_sqnorm([x]), 4.75)
-
-    @attr.gpu
-    def test_sqnorm_scalar_gpu(self):
-        a = cuda.to_gpu(self.a)
-        self.assertAlmostEqual(gradient_clipping._sum_sqnorm([a]), 4)
-
-    @attr.gpu
-    def test_sqnorm_array(self):
-        x = cuda.to_gpu(self.x)
-        a = cuda.to_gpu(self.a)
-        self.assertAlmostEqual(gradient_clipping._sum_sqnorm(
-            [self.x, self.a, x, a]), 8.75 * 2)
-
-    @attr.multi_gpu(2)
-    def test_sqnorm_array_multi_gpu(self):
-        x0 = cuda.to_gpu(self.x, device=0)
-        x1 = cuda.to_gpu(self.x, device=1)
-        a0 = cuda.to_gpu(self.a, device=0)
-        a1 = cuda.to_gpu(self.a, device=1)
-        self.assertAlmostEqual(gradient_clipping._sum_sqnorm(
-            [self.x, self.a, x0, a0, x1, a1]), 8.75 * 3)
+            for i, p in enumerate(params):
+                setattr(self, 'p{}'.format(i), p)
 
 
+@testing.backend.inject_backend_tests(None, _backend_params)
+@testing.backend.inject_backend_tests(None, _backend_params)
+@testing.backend.inject_backend_tests(None, _backend_params)
 class TestGradientClipping(unittest.TestCase):
 
     def setUp(self):
-        self.target = SimpleLink(
-            np.arange(6, dtype=np.float32).reshape(2, 3),
-            np.arange(3, -3, -1, dtype=np.float32).reshape(2, 3))
-
-    def check_clipping(self, multiplier):
-        w = self.target.param.data
-        g = self.target.param.grad
-        xp = backend.get_array_module(w)
+        num_params = 3
+        arrs = [
+            np.random.uniform(-3, 3, (2, 3)).astype(np.float32)
+            for _ in range(num_params)]
+        grads = [
+            np.random.uniform(-3, 3, (2, 3)).astype(np.float32)
+            for _ in range(num_params)]
+        params = []
+        for arr, grad in zip(arrs, grads):
+            param = chainer.Parameter(arr)
+            param.grad = grad
+            params.append(param)
+
+        self.target = SimpleLink(params)
+        self.norm = math.sqrt(sum([g.ravel().dot(g.ravel()) for g in grads]))
+
+    def check_clipping(self, backend_configs, multiplier):
```

Comment:
How about using "rate" instead of "multiplier" to make the correspondence with the source code clearer?

Reply:
Fixed.